pageserver/src/tenant/timeline.rs

```diff
@@ -2709,28 +2713,45 @@ impl Timeline {
                         assert!(!*expect_initdb_optimization, "expected initdb optimization");
                     }
                 }
-                // normal case, write out a L0 delta layer file.
-                let (delta_path, metadata) = self.create_delta_layer(&frozen_layer).await?;
-                HashMap::from([(delta_path, metadata)])
+                // Normal case, write out a L0 delta layer file.
+                // For delta layers, `create_delta_layer` will not modify the layer map.
+                // We will remove frozen layer and add delta layer in one atomic operation later.
+                let layer = self.create_delta_layer(&frozen_layer).await?;
+                (
+                    HashMap::from([(layer.filename(), LayerFileMetadata::new(layer.file_size()))]),
+                    Some(layer),
+                )
             };
 
-        pausable_failpoint!("flush-frozen-before-sync");
-
         // The new on-disk layers are now in the layer map. We can remove the
         // in-memory layer from the map now. The flushed layer is stored in
         // the mapping in `create_delta_layer`.
         {
             let mut guard = self.layers.write().await;
-            let l = guard.layer_map_mut().frozen_layers.pop_front();
 
-            // Only one thread may call this function at a time (for this
-            // timeline). If two threads tried to flush the same frozen
-            // layer to disk at the same time, that would not work.
-            assert!(compare_arced_layers(&l.unwrap(), &frozen_layer));
+            if let Some(ref l) = delta_layer_to_add {
```

Comment:
This could be moved to the L2722R2729 line where the `{` block opens (setting the lifetime for `guard`).

As in we don't need the write lock if we are doing the "initdb image layer" trick.

Though.. One could ask why don't we have completly distinct paths for these.

Reply:
it seems that we need `guard` when `record_residence_event`? (Though I'm not sure why we have to hold the write lock when doing this...)

Reply:
> (Though I'm not sure why we have to hold the write lock when doing this...)

Yes even I've been wondering the same on some bugfix I've been doing ... it's a witness for exposing the layer at the same locking round. Thought it should had been a write lock guard, it was `&_` last I looked.

Reply:
Aah yes now I understand when in files view, oki, good to resolve.